### PR TITLE
Create `nft.approvals`

### PR DESCRIPTION
**Pull request by hildobby**
Brief comments on the purpose of your changes:

Abstract all nft event approval tables (`erc721_{blockchain}.evt_Approval`, `erc721_{blockchain}.evt_ApprovalForAll` & `erc1155_{blockchain}.evt_ApprovalForAll`) for every chain into `nft_{blockchain}.approvals`, before aggregating it all into `nft.approvals`.

Waiting for #3031 to be merged.

I've checked that:

### General checks:
* [ ] I tested the query on dune.com after compiling the model with dbt compile (compiled queries are written to the target directory)
* [ ] I used "refs" to reference other models in this repo and "sources" to reference raw or decoded tables 
* [ ] if adding a new model, I added a test
* [ ] the filename is unique and ends with .sql
* [ ] each sql file is a select statement and has only one view, table or function defined  
* [ ] column names are `lowercase_snake_cased`
* [ ] if adding a new model, I edited the dbt project YAML file with new directory path for both models and seeds (if applicable)
* [ ] if wanting to expose a model in the UI (Dune data explorer), I added a post-hook in the JINJA config to add metadata (blockchains, sector/project, name and contributor Dune usernames)

### Pricing checks:
* [ ] `coin_id` represents the ID of the coin on coinpaprika.com
* [ ] all the coins are active on coinpaprika.com (please remove inactive ones)

### Join logic:
* [ ] if joining to base table (i.e. ethereum transactions or traces), I looked to make it an inner join if possible

### Incremental logic:
* [ ] I used is_incremental & not is_incremental jinja block filters on both base tables and decoded tables
  * [ ] where block_time >= date_trunc("day", now() - interval '1 week')
* [ ] if joining to base table (i.e. ethereum transactions or traces), I applied join condition where block_time >= date_trunc("day", now() - interval '1 week')
* [ ] if joining to prices view, I applied join condition where minute >= date_trunc("day", now() - interval '1 week')

### CODE REVIEW

**Review by 0xRobin — COMMENTED** [review]
Left a couple comments on the arbitrum model (which should apply to all of them).
Mostly on simplifying the table design since this will be a new sector table. 

I'm not certain if we should materialize these models since it will just duplicate all the data in the approval events tables? (not saying we shouldn't, just not sure..) 
Paging @jeff-dude @augustog for some help on that decision, I know `nft.transfers` models are quite heavy on the backend.

File: models/nft/arbitrum/nft_arbitrum_approvals.sql [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+, app.evt_block_time AS block_time
+, date_trunc('day', app.evt_block_time) AS block_date
+, app.evt_block_number AS block_number
+, app.owner AS address
+, 'erc721' AS token_standard
+, CAST(false AS boolean) AS approval_for_all
+, app.contract_address
+, CAST(app.tokenId AS DECIMAL(38,0)) AS token_id
+, approved 
+, CAST(NULL AS boolean) AS approved_for_all
```
[the remark this answers is not in the document]
Comment (hildobby) [reply models/nft/arbitrum/nft_arbitrum_approvals.sql:20]: yeah very good call!

File: models/nft/arbitrum/nft_arbitrum_approvals.sql [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+, app.evt_tx_hash AS tx_hash
+, et.from AS tx_from
+, et.to AS tx_to
```
[the remark this answers is not in the document]
Comment (hildobby) [reply models/nft/arbitrum/nft_arbitrum_approvals.sql]: Commented those out for now so we can always easily add them later if needed. All use cases I will have will use those columns but I'll do the join in the queries for now

**Comment by hildobby** [comment]
@0xRobin I'm working on a new `attacks` sector (`attacks.all` which will be based on `attacks.phishing`, `attacks.scams`, `attacks.exploits`) that you will soon see in a new PR, it will leverage this PR as approvals play a big part in phishing. I've already got 3 types of phishing attacks all automatically detected in queries and will submit those very soon, I'm pretty excited about this research 🫡

**Comment by 0xRobin** [comment]
@hildobby we should either do a `merge` strategy or change the incremental to 
```
WHERE app.evt_block_time >= (select max(block_time) from {{this}})
```
Right now we're always appending the last week again which leads to the duplicates.

**Comment by hildobby** [comment]
Yeah that's what I thought, I wanted to test if merge was the default and not append, which seems to be the case, so I'll add the incremental_strategy back for clarity. 

With merge, `WHERE app.evt_block_time >= (select max(block_time) from {{this}})` isn't necessary, would it make it more efficient? Hard for me to determine if fetching latest block_time from {{this}} is costly or not @0xRobin

**Comment by 0xRobin** [comment]
> Yeah that's what I thought, I wanted to test if merge was the default and not append, which seems to be the case, so I'll add the incremental_strategy back for clarity.
> 
> With merge, `WHERE app.evt_block_time >= (select max(block_time) from {{this}})` isn't necessary, would it make it more efficient? Hard for me to determine if fetching latest block_time from {{this}} is costly or not @0xRobin

While `max from {{this}} ` should be more efficient in theory, we've seen mixed results in the past so we mainly stick to `now() - 1 week.`

It does seem that `merge` is the default, which is also strange because all docs point to `append` as the default. :thinking:

**Comment by hildobby** [comment]
I think merge, while a bit more costly makes more sense across the board to avoid duplicates.

Let's wait for the tests to run, but I think this PR should be good now btw

**Review by 0xRobin — APPROVED** [review]
@hildobby  yes, looks good :+1:

**Review by jeff-dude — APPROVED** [review]
pushed final tweaks to materialize the `nft` schema spell, similar to recent merge for all spells in this schema. we're looking to improve query performance in the dune app, since these spells are so heavily used. reference [here](https://github.com/duneanalytics/spellbook/pull/3093)

also cleaned up an extra column and removed casting in the final layer.

**Pull request merged**